motionplan/kinematic.go

```diff
@@ -24,7 +25,7 @@ func ComputePosition(model referenceframe.Frame, joints *pb.JointPositions) (spa
 	}
 
 	pose, err := model.Transform(model.InputFromProtobuf(joints))
-	if err != nil {
+	if err != nil && !strings.Contains(err.Error(), referenceframe.OOBErrString) {
```

Comment:
So ComputePosition has been modified to NOT return `nil` if we have an OOB error. Is this what we want to be doing? It feels wrong that we could be computing a position with known bad inputs and returning that pose. There may also be some value in doing so, but I would like us to think about the implications here.

Reply:
If we try to Transform something that is out of bounds we return an error, which is the correct thing to do.  But users of this function can also check to see if the error is an out of bounds error and handle that exception according to how they want

Reply:
Deleted the last comment because I realized that it was not right.  You are correct good callout, we should probably be floating the OOBError higher in the stack because now its being squashed in `frame.go`.  However, this change isn't related to the ticket, so we should probably make another one

Reply:
Agreed; this should live elsewhere.

The current defined behavior for `Transform` on a `Frame` is that if OOB, it will return an Error, *and* the OOB pose.

However, this is not a method on a Frame, but a public method. We shouldn't be changing the behavior here, since it's public. We should re-examine whether we want to use a public method at the point in question, and if so, we should create a new one that explicitly will return OOB positions.

Reply:
The most prominent change that `&& !strings.Contains(err.Error(), referenceframe.OOBErrString)` introduced is that if an arm is initialized OOB, then the arm card in the control tab of app.viam
 is empty. 
<img width="1642" alt="Screen Shot 2023-01-31 at 3 47 06 PM" src="https://user-images.githubusercontent.com/46872531/215879241-cf6fac72-a9ba-47c2-972e-6f8d49927dbc.png">

However, using the provided code in the code sample tab, we are still able to connect to the robot and rescue the arm from the OOB state.

Hence, the addition to the conditional allows us to rescue the from the OOB state through the UI and does not force us to use any "code" for rescue purposes.

Reply:
need to do some thinking abt this.. 🤔

Reply:
ok after a night's rest I am back with some new thoughts.
to-reiterate, part of the ticket description states:
> This should be updated such that if an arm is currently in an out-of-bounds state, it is still able to report its present state

Hence, when in an OOB state, `arm.EndPosition` and `arm.JointPositions` should still return the desired information.

> There may also be some value in continuing with this change, but I would like us to think about the implications here.

The addition of `!strings.Contains(err.Error(), referenceframe.OOBErrString)` to the conditional inside `motionplan/kinematic.go/ComputePosition()` makes it such that `components/arm/arm.go/CreateStatus()` does _not_ return `nil, err` when in an OOB state. 

If we _do_ return `nil, err`, this means that the arm card on the UI is blank. 
<img width="1642" alt="Screen Shot 2023-01-31 at 3 47 06 PM" src="https://user-images.githubusercontent.com/46872531/216107711-e7e92805-32e0-4f73-aee7-46cde77e3164.png">
Hence, to rescue an OOB arm we must rely on the use of the SDK.

In my opinion, the loss of the arm UI card is **detrimental** to the user experience for this specific case. However, this can be fixed through submitting a ticket to fleet mgmt and having a more in depth discussion about how to shape the UI when OOB..

> You are correct good callout, we should probably be floating the OOBError higher in the stack because now its being squashed in frame.go
> We shouldn't be changing the behavior here, since it's public. We should re-examine whether we want to use a public method at the point in question, and if so, we should create a new one that explicitly will return OOB positions.

I agree that we should float the OOBError higher in the stack. Because we "shouldn't be changing the behavior here, since it's public", I propose new public method inside `motionplan/kinematic.go` called `ComputeOOBPositions()`. Then, `ComputePositions()` would error on OOB, while `ComputeOOBPositions()` would not error on OOB. 

This means that all arm driver implementations of `EndPosition()` could be changed from:
```
return motionplan.ComputePosition(e.model, joints)
```
to
```
return motionplan.ComputeOOBPosition(e.model, joints)
```

In turn, all arm driver implementations of `MoveToPosition()` could be changed from:
```
model := e.ModelFrame()
joints, err := e.JointPositions(ctx, nil)
if err != nil {
	return err
}
// check that joint positions are not out of bounds
if _, err = model.Transform(model.InputFromProtobuf(joints)); err != nil {
	return err
}
```
to
```
joints, err := e.JointPositions(ctx, nil)
if err != nil {
	return err
}
if _, err = motionplan.ComputePosition(e.model, joints); err != nil {
	return err
}
```

Reply:
This is great. `ComputeOOBPosition()` is the correct path to go down here. Thus `ComputePosition()` maintains its current behavior, and a user must be explicit about the call if they want OOB to be allowable.

I have one concern here: if we implement the above as-specified, while the arm card will now be valid, if a joint is indeed OOB, then the user needs to know that that is the case, and be told exactly what to do to fix it, since position moves will be failing.

Can we update the error statement to specify exactly which joint is OOB, in addition to the current "val needs to be within bounds [min, max]"?